rfcs/005-metadata-rework.md

```diff
@@ -0,0 +1,562 @@
+
+nickel>let base_hidden = {
+  foo = ((("bar" | doc "Foo") | default) | Str)
+}
+nickel>:query base_hidden.foo
+• contract: Str
+• value: "bar"
+
+```
+
+One solution could be to define a flattening operation (pedantically, a monoid
+operation, or a monad if metadata are rather seen as a container over a value):
+`flatten : (Metadata, Metadata) -> Metadata`.
+
+While fetching metadata can take an arbitrary time to complete _in principle_,
+it's currently quick for the happy path, when metadata are located e.g. directly
+on record fields. Now, if we now want to correctly handle nested metadata, we
+can't stop evaluation as soon as we encounter metadata anymore, because some
+other metadata could be lying underneath. We now have to _systematically
+evaluate an expression down to a value_, collecting and flattening metadata
+along the way.
+
+## Proposal
+
+In a configuration language such as Nickel, the fundamental data structure is
+the record, and the fundamental piece of data is the record field. In
+retrospect, fields seem to be the right level of granularity for metadata, and
+where they are most often attached in practice.
+
+While metadata that can appear anywhere are in theory more flexible, the only
+example making use of such capability of metadata to be attached to other values
+than record fields is the example where metadata are propagated through an array
+and another record field. This example is contrived, and doesn't seem to
+correspond to a real use-case in a convincing way. In the NixOS module system,
+for example, metadata are attached to options only.
+
+Field-based metadata would only be affected during merging, and combined eagerly
+when that happens: no later arbitrary evaluation, no nesting, just static data.
+Allowing them anywhere doesn't bring clear practical benefits, whereas it comes
+with substantial drawbacks, exposed in the previous sections.
+
+### Specification
+
+The proposed change comes in to aspects:
+
+1. Metadata can only be written next to record fields, and are stored only there
+   in the AST. Their semantics with respect to merging is left unchanged, except
+   for cross-application of contracts (see the next point). Contract annotations
+   and type annotations are still allowed on any value, but would be parsed as a
+   different node (say, `Annotated`) and wouldn't impact the behavior of
+   merging.
+2. Instead of implementation `(CrossApp)`, we implement a laxer strategy,
+   lazy propagation (which corresponds closely to `(EvVal)`). As we've done with
+   [lazy array contracts](https://github.com/tweag/nickel/pull/764), contracts
+   would be accumulated at the field level instead of being eagerly applied.
+   Only when the field is extracted (or evaluated by any other mean: equality
+   testing, deep sequing, etc.), we can apply the pending contracts. As for lazy
+   array contracts, this also enables new potential optimizations, like
+   eliminating redundant contract applications.
```

Comment:
We're adding contract conjunction again, aren't we? This may be an issue. We a least need to clarify the semantics here.

Reply:
In some sense, yes, but in fact, we already do this currently (when cross-applying contracts). Here we just propose to do that a bit more lazily. The semantics is that this should be the same as annotating a value with several contracts at once:

```
foo | A
    | B
    | C
```

Which just means applying the contract in sequence, id est the "naive" intersection. Do you think it needs to be emphasized?

Reply:
Yes, it's probably worth it, and it can't hurt.

On function it's more or less `(A → B) ∩ (C → D) = (A∩C) → (B∩D)`, correct?

Reply:
> On functions it's more or less (A → B) ∩ (C → D) = (A∩C) → (B∩D), correct?

Correct.